Fixed on trunk with
http://svn.gnome.org/viewvc/nautilus/trunk/libnautilus-private/nautilus-emblem-utils.c?r1=14319&r2=14357&view=patch
and
http://svn.gnome.org/viewvc/nautilus/trunk/libnautilus-private/nautilus-file.c?r1=14318&r2=14319
so shouldn't be necessary for nautilus-2.24 onwards

--- libnautilus-private/nautilus-emblem-utils.c.orig	2008-09-28 11:40:05.284668861 +0300
+++ libnautilus-private/nautilus-emblem-utils.c	2008-09-28 11:40:12.617466465 +0300
@@ -30,6 +30,8 @@
 #include <string.h>
 #include <time.h>
 #include <unistd.h>
+#include <sys/stat.h>
+
 #include "nautilus-file.h"
 #include <eel/eel-glib-extensions.h>
 #include <eel/eel-gdk-pixbuf-extensions.h>
--- libnautilus-private/nautilus-file.c.orig	2008-09-28 11:49:33.990743377 +0300
+++ libnautilus-private/nautilus-file.c	2008-09-28 11:50:20.917404516 +0300
@@ -56,6 +56,7 @@
 #include <grp.h>
 #include <gtk/gtksignal.h>
 #include <glib/gi18n.h>
+#include <glib/gstdio.h>
 #include <gio/gio.h>
 #include <glib/gurifuncs.h>
 #include <libgnome/gnome-macros.h>
